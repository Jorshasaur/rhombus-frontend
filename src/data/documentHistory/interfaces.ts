export interface DocumentHistoryState {
    showDocumentHistory: boolean
}
