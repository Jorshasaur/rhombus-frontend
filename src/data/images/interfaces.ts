export interface ActiveImageInfo {
    fileName: string
    author: string
    createdAt?: string
    url?: string
    width?: number
    height?: number
}

export interface ImageCarouselImage {
    url: string
    fileName: string
    width?: number
    height?: number
}

export interface ImageCarouselData {
    left: ImageCarouselImage
    middle: ImageCarouselImage
    right: ImageCarouselImage
}

export interface ImageDataStore {
    id: string
    assetId?: string
    authorId: string
    url: string
    createdAt?: string
    width?: number
    height?: number
}

export interface ImagesStoreState {
    activeImage: string
    imageDatas: { [imageId: string]: ImageDataStore }
    imageList: string[]
}
