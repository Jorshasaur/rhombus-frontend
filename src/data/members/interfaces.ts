export interface NavigationMember {
    avatarURL: string
    id: number
    index: number
    name: string
    useAvatarURL: boolean
}
