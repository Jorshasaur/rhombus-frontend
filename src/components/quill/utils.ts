import { DeltaStatic } from 'quill-delta'
import Quill from 'quill/core'
import { CommentMention } from '../../data/services/PagesApiService'
import { Blot } from 'parchment/dist/src/blot/abstract/blot'

interface CommentTextAndMentions {
    text: string
    source: string
    mentions: CommentMention[]
}

export function getBounds(bounds: ClientRect, containerBounds: ClientRect) {
    return {
        top: bounds.top - containerBounds.top,
        right: bounds.right - containerBounds.left,
        bottom: bounds.bottom - containerBounds.top,
        left: bounds.left - containerBounds.left,
        width: bounds.width,
        height: bounds.height
    }
}

export function getCachedContents(quill: Quill) {
    return quill.editor.delta
}

export function getTextAndMentions(delta: DeltaStatic) {
    const res = { text: '', source: '', mentions: [] }

    if (!delta) {
        return res
    }

    return delta.reduce((ret: CommentTextAndMentions, op) => {
        if (op.insert != null) {
            if (typeof op.insert === 'string') {
                ret.text += op.insert
                ret.source += op.insert
            } else if (op.insert.mention) {
                const { name, userId } = op.insert.mention
                ret.text += `@${name}`
                ret.source += `<@U${userId}>`
                ret.mentions.push({
                    token: name,
                    userId
                })
            }
        }
        return ret
    }, res)
}

export function preloadImage(imageUrl: string) {
    const preload = new Image()
    preload.src = imageUrl
}

export function getBlotOffset(quill: Quill, blot: Blot) {
    let offset = blot.offset()

    while (blot.parent !== quill.scroll) {
        blot = blot.parent
        offset += blot.offset()
    }
    return offset
}

// It's odd but the line length for the added last line is actually 1 because of the
// invisible line break character added to create the line.
export function isBlankLine(lineLength: number, lineContent: string) {
    return lineLength === 1 && lineContent === ''
}
