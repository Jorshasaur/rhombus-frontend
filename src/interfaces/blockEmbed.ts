import FileEmbed from '../components/pages/Editor/Blots/FileEmbed'
import DefaultEmbedTypes from '../components/quill/blots/DefaultEmbedTypes'
import { BlotSize } from './blotSize'

export type BlockEmbedService =
    | keyof typeof DefaultEmbedTypes
    | ReturnType<FileEmbed['_getFileTypeClass']>

export interface BlockEmbedValue {
    authorId: string
    createdAt: string
    dataUrl?: string
    embed?: any
    embedData: any
    originalLink?: string
    service: BlockEmbedService
    size?: BlotSize
    type?: string
    unviewable?: boolean
    uuid: string
    version: number
}

export interface BlockEmbedProps {
    authorId: string
    authorName?: string
    createdAt?: string
    embedData?: {}
    key: string
    originalLink?: string
    service: BlockEmbedService
    size?: BlotSize
    title?: string
    type?: string
    uuid?: string
    version?: number
}

export interface ResizableEmbed {
    _renderSmall: Function
    _renderFullSize: Function
    handleModalOpen: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void
    handleModalClose: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void
}
