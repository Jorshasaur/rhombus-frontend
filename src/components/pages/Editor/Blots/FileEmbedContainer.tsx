import urlModule from 'url'
import PagesApiService from '../../../../data/services/PagesApiService'
import store from '../../../../data/store'
import tempAssets from '../../../../data/tempAssets'
import { getAsset } from '../../../../data/utils/assets'
import { Asset } from '../../../../interfaces/asset'
import { isAssetUrl } from '../../../../lib/utils'
import EmbedContainer, { EmbedContainerAppState } from './EmbedContainer'

export interface FileEmbedContainerAppState extends EmbedContainerAppState {
    dataUrl?: string
    asset?: Asset
}

export default class FileEmbedContainer extends EmbedContainer {
    state: FileEmbedContainerAppState

    async receivedNewState() {
        const { embedData, asset, uuid } = this.state
        if (asset == null && embedData != null && embedData.url == null) {
            if (embedData.id != null) {
                let newAsset = getAsset(store, embedData.id) as
                    | Asset
                    | undefined
                if (newAsset == null) {
                    newAsset = await PagesApiService.getAsset(embedData.id)
                    if (newAsset == null || newAsset.id == null) {
                        await this.copyAsset(embedData.id)
                        return
                    }
                }
                this.setState({ asset: newAsset })
            } else if (this.state.dataUrl == null) {
                const tempAsset = tempAssets.getAsset(uuid)
                if (tempAsset != null && isAssetUrl(tempAsset.url)) {
                    await this.copyAssetFromUrl(tempAsset.url)
                }
            }
        }
    }

    async copyAssetFromUrl(assetUrl: string) {
        const assetUrlPath = urlModule
            .parse(assetUrl)
            .path!.replace('/assets/', '')
        const asset = await PagesApiService.copyAssetFromUrl(assetUrlPath)
        if (asset != null && asset.id) {
            this.setState({ asset: asset })
            this.setEmbedDataValue('id', asset.id)
        }
    }

    async copyAsset(assetId: string) {
        const reduxState = store.getState()
        if (reduxState.currentDocument.id) {
            const newAsset = await PagesApiService.copyAsset(assetId)
            if (newAsset != null && newAsset.id) {
                this.setState({ asset: newAsset })
                this.setEmbedDataValue('id', newAsset.id)
            }
        }
    }
}
