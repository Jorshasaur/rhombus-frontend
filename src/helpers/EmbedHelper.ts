import ReactDOM from 'react-dom'
import { Maybe } from 'true-myth'
import URI from 'urijs'
import { BlotSize } from '../interfaces/blotSize'
import {
    DEFAULT_RATIO,
    GUTTER_WIDTH,
    LARGE_IMAGE_SIDE_MARGIN,
    MEDIUM_BREAKPOINT,
    SMALL_BREAKPOINT
} from '../constants/styles'
import { getFileExtension } from '../lib/utils'

export enum FILE_CREATE_METHOD {
    plusButton = 'PLUS_BUTTON',
    fileDrop = 'FILE_DROP',
    paste = 'PASTE',
    other = 'OTHER'
}

export enum FILE_TYPES {
    mov = 'video/quicktime',
    mp4 = 'video/mp4',
    pdf = 'application/pdf',
    png = 'image/png',
    jpeg = 'image/jpeg',
    jpg = 'image/jpeg',
    sketch = 'application/sketch',
    studio = 'application/studio'
}

export enum URL_TYPES {
    freehand = 'freehand',
    prototype = 'prototype',
    invision = 'invision',
    externalURL = 'externalUrl'
}

export function getEmbedType(fileType: string, fileName: string): string {
    if (fileType === '') {
        return FILE_TYPES[getFileExtension(fileName) || 'notthere']
    }

    return fileType
}

export function getLargeEmbedWidth() {
    return window.innerWidth - GUTTER_WIDTH * 2 - LARGE_IMAGE_SIDE_MARGIN * 2
}

export function getContainerWidth(container: HTMLElement) {
    const computedStyle = window.getComputedStyle(container)
    let containerWidth = container.getBoundingClientRect().width
    containerWidth -=
        parseFloat(computedStyle.paddingLeft || '0') +
        parseFloat(computedStyle.paddingRight || '0')
    return containerWidth
}

export function getLargeEmbedLeft(
    containerWidth: number,
    mediumSizeAdjust: number = 0
) {
    if (window.innerWidth <= SMALL_BREAKPOINT) {
        return 0
    } else if (window.innerWidth <= MEDIUM_BREAKPOINT) {
        const columnWidth = containerWidth / 7
        const halfGutter = GUTTER_WIDTH / 2
        return (
            (containerWidth - getLargeEmbedWidth()) / 2 +
            columnWidth -
            halfGutter * 3 +
            mediumSizeAdjust
        )
    }
    return (containerWidth - getLargeEmbedWidth()) / 2
}

export const getEmbedStyles = (
    aspectRatio: number,
    container: HTMLElement,
    size: BlotSize,
    hasOpenThread: boolean = false
): {
    width: number
    height: number
    left: number
} => {
    const containerWidth = getContainerWidth(container)
    const style = {
        left: 0,
        width: containerWidth,
        height: containerWidth * aspectRatio
    }
    if (size === BlotSize.Large && window.innerWidth > SMALL_BREAKPOINT) {
        const largeEmbedWidth = getLargeEmbedWidth()
        const largeEmbedLeft = getLargeEmbedLeft(containerWidth)
        const threadWidth = containerWidth + Math.abs(largeEmbedLeft)
        return {
            height:
                aspectRatio * (hasOpenThread ? threadWidth : largeEmbedWidth),
            left: largeEmbedLeft,
            width: hasOpenThread ? threadWidth : largeEmbedWidth
        }
    }

    return style
}

export const getEmbedSize = <T>(
    embed: React.Component<T>,
    extendedInfoSelector: string
) => {
    const defaultHeight =
        (document.querySelector('.ql-editor')!.clientWidth * 10) / 16
    return Maybe.of(ReactDOM.findDOMNode(embed) as Element)
        .andThen((node) =>
            Maybe.of(node.querySelector(`.${extendedInfoSelector}`))
        )
        .map((extendedInfo) => ({
            height: extendedInfo.clientHeight,
            width: extendedInfo.clientWidth
        }))
        .unwrapOr({
            height: defaultHeight,
            width: defaultHeight * DEFAULT_RATIO
        })
}

export function isOnTeam(uri: ReturnType<typeof URI>): boolean {
    if (process.env.NODE_ENV === 'development') {
        return true
    }
    const currentUri = new URI()
    return uri.subdomain() === currentUri.subdomain()
}

const PLAY_SEGMENT = '/play/'
const INSPECT_SEGMENT = '/inspect/'
const MOTION_SEGMENT = /\/motion.*/gi
const COMMENT_SEGMENT = '/comment/'

export function getPlayUrl(originalLink: string) {
    let link = originalLink

    if (link.toLowerCase().indexOf(INSPECT_SEGMENT) > -1) {
        link = link
            .replace(new RegExp(INSPECT_SEGMENT, 'ig'), PLAY_SEGMENT)
            .replace(MOTION_SEGMENT, '')
    } else if (link.toLowerCase().indexOf(COMMENT_SEGMENT) > -1) {
        link = link.replace(new RegExp(COMMENT_SEGMENT, 'ig'), PLAY_SEGMENT)
    }

    const uri = URI(link)
    link = uri.removeSearch('hideNavBar').toString()

    return link
}
