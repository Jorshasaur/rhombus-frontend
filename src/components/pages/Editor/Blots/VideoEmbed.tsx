import React from 'react'
import { Subscribe } from 'unstated'
import cx from 'classnames'
import { BlockEmbedProps } from '../../../../interfaces/blockEmbed'
import FileEmbedContainer from './FileEmbedContainer'
import { AnimatedEmbedWrapper } from './AnimatedEmbedWrapper'
import styles from './Blots.module.css'
import { get } from 'lodash'
import { BlotSize } from '../../../../interfaces/blotSize'
import { getEmbedSize } from '../../../../helpers/EmbedHelper'
import { PersistentBar } from './PersistentBar'
import { Timestamp } from './Timestamp'
import { Asset } from '../../../../interfaces/asset'
import { Skeleton } from '@invisionapp/helios'
import { DEFAULT_RATIO } from '../../../../constants/styles'
import videoJS from 'video.js'
import { improveVideoScrubbing } from '../../../../lib/utils'
import '!style-loader!css-loader!video.js/dist/video-js.css'
import './Video.css'

interface Props extends BlockEmbedProps {
    container: HTMLElement
}

interface VideoProps {
    asset: Asset
}

class Video extends React.Component<VideoProps> {
    videoRef = React.createRef<HTMLVideoElement>()

    componentDidMount() {
        if (this.videoRef.current != null) {
            improveVideoScrubbing()
            const video = videoJS(this.videoRef.current, {
                fluid: true,
                liveui: true
            })
            video.aspectRatio('16:9')
        }
    }

    getContentType(asset: Asset) {
        if (asset.contentType === 'video/quicktime') {
            return 'video/mp4'
        } else {
            return asset.contentType
        }
    }

    render() {
        const { asset } = this.props

        return (
            <React.Fragment>
                <div className={styles.extendedInfo}>
                    <video
                        controls
                        className={cx(
                            'video-js',
                            'vjs-big-play-centered',
                            styles.video
                        )}
                        ref={this.videoRef}>
                        <source
                            src={`${asset.url}?filename=${asset.fileName}`}
                            type={this.getContentType(asset)}
                        />
                    </video>
                </div>
            </React.Fragment>
        )
    }
}

export default class VideoEmbed extends React.Component<Props> {
    _renderPlaceholder = () => {
        return (
            <div className={styles.extendedInfo}>
                <Skeleton ratio={DEFAULT_RATIO} />
            </div>
        )
    }

    _renderFullSize = (embed: FileEmbedContainer) => {
        const { asset, embedData } = embed.state
        return (
            <React.Fragment>
                <div
                    className={cx(
                        styles.blockEmbedWrapper,
                        this.props.service,
                        this.props.uuid
                    )}>
                    {asset ? (
                        <Video asset={asset} />
                    ) : (
                        this._renderPlaceholder()
                    )}
                </div>
                <PersistentBar
                    service={this.props.service}
                    originalLink={this.props.originalLink}
                    name={get(asset, 'fileName') || embedData.fileName}
                    hoverChildren={
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <Timestamp
                                author={this.props.authorName}
                                className={styles.persistentTimestamp}
                                createdAt={this.props.createdAt}
                            />
                        </div>
                    }
                />
            </React.Fragment>
        )
    }

    render() {
        return (
            <Subscribe to={[FileEmbedContainer]}>
                {(embed: FileEmbedContainer) => {
                    return (
                        <AnimatedEmbedWrapper
                            renderFullSize={() => this._renderFullSize(embed)}
                            size={embed.state.size || BlotSize.Medium}
                            container={this.props.container}
                            getEmbedSize={() =>
                                getEmbedSize(this, styles.extendedInfo)
                            }
                            hasOpenThread={embed.state.hasOpenThread || false}
                        />
                    )
                }}
            </Subscribe>
        )
    }
}
