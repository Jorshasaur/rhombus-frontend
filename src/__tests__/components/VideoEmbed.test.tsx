import React from 'react'
import { mount } from 'enzyme'
import { ThemeProvider } from 'styled-components'
import theme from '@invisionapp/helios/css/theme'
import { Provider } from 'unstated'
import { BlockEmbedService } from '../../interfaces/blockEmbed'
import { createContainer, setupQlEditor } from '../utils'
import { BlotSize } from '../../interfaces/blotSize'
import * as EmbedHelpers from '../../helpers/EmbedHelper'
import FileEmbedContainer from '../../components/pages/Editor/Blots/FileEmbedContainer'
import VideoEmbed from '../../components/pages/Editor/Blots/VideoEmbed'
import MockDate from 'mockdate'

const asset = {
    id: '123',
    url: 'http://video.mp4',
    fileName: 'test.mp4'
}

jest.mock('../../data/store', () => {
    return {
        subscribe: jest.fn(),
        getState: jest.fn(() => ({
            permissions: {
                canEdit: true
            },
            assets: {
                '123': asset
            }
        }))
    }
})

const Component = ({
    children,
    provider
}: {
    children: React.ReactNode
    provider: any
}) => (
    <ThemeProvider theme={theme}>
        <Provider inject={[provider]}>{children}</Provider>
    </ThemeProvider>
)

const service: BlockEmbedService = 'video'

const baseProps = {
    authorId: '1',
    authorName: 'Bob the Builder',
    key: '123456',
    service,
    uuid: '123456',
    version: 1,
    embedData: { id: '123', fileName: 'test.mp4' },
    container: createContainer(100, 100),
    size: BlotSize.Medium,
    createdAt: '2018-04-24T20:09:17.175Z'
}

describe('VideoEmbed component', () => {
    beforeAll(() => {
        setupQlEditor()
        jest.spyOn(EmbedHelpers, 'getContainerWidth').mockReturnValue(100)
        jest.spyOn(EmbedHelpers, 'getEmbedSize').mockReturnValue({
            height: 100,
            width: 100
        })
        window.getComputedStyle = jest.fn(() => {
            return {}
        })
    })

    it('should render a video embed', () => {
        MockDate.set('2019-04-24T20:09:17.175Z')
        const provider = new FileEmbedContainer(baseProps)

        const wrapper = mount(
            <Component provider={provider}>
                <VideoEmbed {...baseProps} />
            </Component>
        )

        const video = wrapper.find('VideoEmbed')
        expect(video).toMatchSnapshot()
    })

    it('should render a placeholder when there is no asset', () => {
        const props = {
            ...baseProps,
            embedData: { id: '222', fileName: 'placeholder.mp4' }
        }

        const provider = new FileEmbedContainer(props)

        const wrapper = mount(
            <Component provider={provider}>
                <VideoEmbed {...props} />
            </Component>
        )

        expect(wrapper.find('Skeleton')).toHaveLength(1)
    })
})
