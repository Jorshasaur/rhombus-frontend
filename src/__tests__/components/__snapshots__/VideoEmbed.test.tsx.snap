// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoEmbed component should render a video embed 1`] = `
<VideoEmbed
  authorId="1"
  authorName="Bob the Builder"
  container={<div />}
  createdAt="2018-04-24T20:09:17.175Z"
  embedData={
    Object {
      "fileName": "test.mp4",
      "id": "123",
    }
  }
  key="123456"
  service="video"
  size="medium"
  uuid="123456"
  version={1}
>
  <Subscribe
    to={
      Array [
        [Function],
      ]
    }
  >
    <Consumer>
      <AnimatedEmbedWrapper
        container={<div />}
        getEmbedSize={[Function]}
        hasOpenThread={false}
        onTransitionEnd={[Function]}
        renderFullSize={[Function]}
        size="medium"
      >
        <div
          className="container"
          data-testid="animated-embed-wrapper__medium"
          onTransitionEnd={[Function]}
          style={
            Object {
              "height": 140,
              "left": 0,
              "overflow": "hidden",
              "width": 100,
            }
          }
        >
          <div
            className="medium"
            style={
              Object {
                "opacity": 1,
                "pointerEvents": "unset",
              }
            }
          >
            <div
              className="blockEmbedWrapper video 123456"
            >
              <Video
                asset={
                  Object {
                    "fileName": "test.mp4",
                    "id": "123",
                    "url": "http://video.mp4",
                  }
                }
              >
                <div
                  className="extendedInfo"
                >
                  <video
                    className="video-js vjs-big-play-centered video"
                    controls={true}
                  >
                    <source
                      src="http://video.mp4?filename=test.mp4"
                    />
                  </video>
                </div>
              </Video>
            </div>
            <PersistentBar
              hoverChildren={
                <div
                  style={
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                    }
                  }
                >
                  <Unknown
                    author="Bob the Builder"
                    className="persistentTimestamp"
                    createdAt="2018-04-24T20:09:17.175Z"
                  />
                </div>
              }
              name="test.mp4"
              service="video"
            >
              <div
                className="persistentBar"
                data-allow-propagation="true"
                data-persistent-bar={true}
                onClick={[Function]}
              >
                <a
                  className="serviceIcon video"
                  data-allow-propagation="true"
                >
                  <svg
                    fill="var(--color-persistent-bar-icon)"
                    height={18}
                    width={21}
                  />
                </a>
                <div
                  className="textContent"
                  data-testid="persistent-bar-title"
                >
                  <div
                    className="title"
                  >
                    test.mp4
                  </div>
                </div>
                <div
                  className="barChildren persistent-bar__children"
                  data-testid="persistent-bar__children"
                />
                <div
                  className="hoverChildren persistent-bar__hover-children"
                  data-testid="persistent-bar__hover-children"
                  onClick={[Function]}
                >
                  <div
                    style={
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                      }
                    }
                  >
                    <Component
                      author="Bob the Builder"
                      className="persistentTimestamp"
                      createdAt="2018-04-24T20:09:17.175Z"
                    >
                      <div
                        className="persistentTimestamp"
                        style={
                          Object {
                            "maxWidth": undefined,
                          }
                        }
                      >
                        Added 
                        <TimeAgo
                          component="time"
                          date="2018-04-24T20:09:17.175Z"
                          formatter={[Function]}
                          live={true}
                          maxPeriod={Infinity}
                          minPeriod={0}
                          now={[Function]}
                        >
                          <time
                            dateTime="2018-04-24T20:09:17.175Z"
                            title="2018-04-24T20:09:17.175Z"
                          >
                            1y ago
                          </time>
                        </TimeAgo>
                         by Bob the Builder
                      </div>
                    </Component>
                  </div>
                </div>
              </div>
            </PersistentBar>
          </div>
        </div>
      </AnimatedEmbedWrapper>
    </Consumer>
  </Subscribe>
</VideoEmbed>
`;
